Feature: Adds new method to extensionRegistry for appendCondition & appendConditions

warrenbuckley opened the pull request:
## Description

Adds two new methods to extensionRegistry to allow developers to append new conditions to an existing extension.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

## Motivation and context
Its not possible to update an extension such as Workspace Entity Action with a new condition. The current approach requires you to overwrite the extension. Which involves setting the same manifest meta properties and reimplementing an API to perform the same action as the core implementation of the button.

**Scenario:**
I want to show/hide the Save and Publish button based on a new condition and keep the existing functionality of the button without overhead of trying to overwrite the extension and keeping it in sync with the core implemetation.


## How to test?
* Read over the PR
* Read the associated tests
* Run the tests - ensure it passes

## Usage
```typescript
// Add a condition with a specific config to Section2
const workspaceCondition:WorkspaceAliasConditionConfig = {
	alias: 'Umb.Condition.WorkspaceAlias',
	match: 'Umb.Workspace.Document'
};
extensionRegistry.addCondition('Umb.Test.Section.2', workspaceCondition);
```

```typescript
const conditions:Array<UmbConditionConfigBase> = [
	{
		alias: 'Umb.Test.Condition.Valid'
	},
	{
		alias: 'Umb.Condition.WorkspaceAlias',
		match: 'Umb.Workspace.Document'
	} as WorkspaceAliasConditionConfig
]

extensionRegistry.addConditions('Umb.Test.Section.1', conditions);
```

## Checklist
- [ ] If my change requires a change to the documentation, I have updated the documentation in this pull request.
- [x] I have read the **[CONTRIBUTING](<(https://github.com/umbraco/Umbraco.CMS.Backoffice/blob/main/.github/CONTRIBUTING.md)>)** document.
- [x] I have added tests to cover my changes.

warrenbuckley committed:
Adds new method to extensionRegistry for appendCondition

warrenbuckley committed:
Adds appendConditions to allow adding multiple in one go

warrenbuckley commented:
Not touched the area with the failing test, so unsure what's going on there...

warrenbuckley commented:
So running the tests from the CodeSpace all pass just fine
<img width="1920" alt="image" src="https://github.com/user-attachments/assets/d709b956-ebb5-4c8f-9d91-50e98b1f7b4f">

warrenbuckley commented:
Thanks for re-running the tests for me @madsrasmussen 
Now I will wait patiently until this gets reviewed or ask for any rework from me

nielslyngsoe commented on src/libs/extension-api/registry/extension.registry.ts:
```diff
@@ -430,4 +430,40 @@ export class UmbExtensionRegistry<
+		const extensionToUpdate = allExtensions.find((ext) => ext.alias === alias) as ManifestWithDynamicConditions;
+
+		if(extensionToUpdate === undefined) {
+			console.error(`Extension with alias ${alias} not found`);
```
I guess this solution could be fine, but ideally the code support late coming Extensions.

So the appending awaits the existence of the extension.

nielslyngsoe replied:
I think this could be done just by making this method async, and then await a method that observes for a specific alias and resolves once its present. A method like this `whenAlias(alias: string): Promise<ManifestTypes>`

warrenbuckley replied:
@nielslyngsoe I will look into this Thurs/Fri when I switch gears for different clients

warrenbuckley replied:
@nielslyngsoe there is a new commit where I have tried to implement this from your notes - hence the highlighted code is outdated

nielslyngsoe commented on src/libs/extension-api/registry/extension.registry.test.ts:
```diff
@@ -453,3 +454,103 @@ describe('UmbExtensionRegistry with exclusions', () => {
 	});
 });
+
+describe('Append Conditions', () => {
```
Super cool with unit tests!

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Adds prependCondition and prependConditions so you can make sure your condition loads first, as order of conditions are important

warrenbuckley committed:
Adds _whenExtensionAliasIsRegistered as a promise so that we basically wait until the manifest is registered so late extensions will work as well

Seems to work with my hack in an entrypoint and the console logs the extension with the updated condition but its the tests I am struggling to get to work.

As this code is heavily worked from me trying to Google, read RXJS docs and talking with Copilot I am not fully in confident in this approach

madsrasmussen committed:
simplify _whenExtensionAliasIsRegistered

madsrasmussen committed:
split some tests

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Rework of tests to pass and minor fix to _whenExtensionAliasIsRegistered to get the ext alias that matches

warrenbuckley committed:
Trying to add a test to show that a late registered extension works - but test says otherwise compared to using it in entrypoint

warrenbuckley commented:
Needing to take a break from this test that is failing to demonstrate late loaded extension.
Unsure if its my code or just my test implemtation at this point.

**Could do with some cleverer people than me from HQ to give me some input/thoughts**
@nielslyngsoe @madsrasmussen or @iOvergaard & others...

```
 🚧 Browser logs:
      About to go KABOOM..

 ❌ Prepend Conditions > allows conditions to be prepended when an extension is loaded later on
      Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
        at createTimeoutError (node_modules/@web/test-runner-mocha/dist/autorun.js:1:191490)
        at ds._timeoutError (node_modules/@web/test-runner-mocha/dist/autorun.js:1:195723)
        at node_modules/@web/test-runner-mocha/dist/autorun.js:1:193804
```

nielslyngsoe commented:
Hi @warrenbuckley you should not be needing to prepend or append. It will give the same result. Currently the system spins up all conditions, the order means nothing. :-)

warrenbuckley commented:
@nielslyngsoe I thought order would be important with last in winning surely?

* Condition A - ✅
* Condition B - ❌
* Condition C - ❌

For me I would have thought that the last one setting permitted to false I accept the extension to be loaded so it overrides the previous ones before it. 

Or with this way it would prevent the extension as last one is false
* Condition A - ❌
* Condition B - ✅
* Condition C - ❌


or perhaps this I would expect condition C to then make the extension appear
* Condition A - ❌
* Condition B - ❌
* Condition C - ✅

So can you confirm how it works please @nielslyngsoe 

Then happy to amend this PR once I fully grasp how conditions work.

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

nielslyngsoe commented:
@warrenbuckley It works in the way that all the listed conditions has to be valid for the extension to be shown. Therefore the other doesn't matter. :-)

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Remove the prepend conditions and rename append to addCondition and addCondtions

As the order of conditions is not important, as all conditions set on an extensions must be true/valid

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Update tests

Still need to uncomment & reinvestigate the late loading test

warrenbuckley committed:
Fix up code so test passes & my rough entrypoint demo of usage works

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

nielslyngsoe committed:
Merge branch 'main' into feature/extRegistry-appendCondition

warrenbuckley committed:
Merge branch 'main' into feature/extRegistry-appendCondition

nielslyngsoe commented on src/libs/extension-api/registry/extension.registry.ts:
```diff
@@ -430,4 +443,62 @@ export class UmbExtensionRegistry<
+	 * @param alias {string} - The alias of the extension to wait for.
+	 * @returns {Promise<ManifestTypes>} - A promise that resolves with the extension.
+	 */
+	private async _whenExtensionAliasIsRegistered(alias: string): Promise<ManifestBase> {
```
Interesting, I will have to test this approach further :-)

nielslyngsoe committed:
Merge branch 'main' into feature/extRegistry-appendCondition

nielslyngsoe committed:
Merge remote-tracking branch 'origin/v15/dev' into feature/extRegistry-appendCondition

nielslyngsoe committed:
import type

nielslyngsoe committed:
remove test code

nielslyngsoe committed:
refactor for a more direct appending implementation

nielslyngsoe commented:
Hi @warrenbuckley 

I have refactored your code and written a few extra tests. I'm happy with the solution now.

Notice I went away for the Async approach as it turned out that appended conditions, would be appended late, like two JS cycles, from when a extension manifest got registered.

In other words, a late coming manifest, would be broadcasted and utilized by all subscribers, for then 2 cycles later to get the additional conditions appended. This was a unnecessary round trip, so I refactored so the additional conditions get appended before the manifest goes into the Extension Registry.

I will wait for the v15/dev branch to get up to date with v.14 before we merge. As you can see there is too many changes as part of this PR right now.

nielslyngsoe committed:
Merge branch 'v15/dev' into feature/extRegistry-appendCondition

nielslyngsoe reviewed (approved):
I'm happy now :-)

warrenbuckley commented:
WHOOP WHOOP 🎉 
Off to read the rework you done and see what I could have done better

warrenbuckley commented:
Just so I am clear, this is for V15 only @nielslyngsoe ?
